### PR TITLE
Add: toggle dark/night mode styles automatically

### LINKED ISSUE #51: Scheduling / night mode
As already suggested ages ago by someone (https://github.com/stylish-userstyles/stylish/issues/242), I would love to have a night mode option, which automatically turns websites dark at specific hours.

It could work by selecting specific styles, which one wants to have automatically activated at a configurable time frame.

Since I mainly use Stylus for that reason (apart from a few very good light themes) this feature would be perfect! :blush:

### LINKED ISSUE #524: Toggle styles according to prefers-color-scheme
As you might know, the newly released macOS Mojave has a great Dark Mode implementation where all supported apps automatically change color theme to Dark when Dark Mode is toggled.

It would be great if Stylus could be compatible with Dark Mode and automatically be toggled based on the Dark Mode status.

### PR DESCRIPTION
**Pull request by eight04**
Fixes #524
Fixes #62 
Fixes #51 

Todos:

* [ ] Options page for `schemeSwitcher.*`. https://github.com/openstyles/stylus/pull/736#issuecomment-541276743
* [x] Add an indicator in the popup if the style is excluded because of unmatched dark/light mode.

Note that we don't have a style setting page yet, so you can only configure this option via the installation page of usercss.

### CODE REVIEW

**Comment by eight04** [comment]
Or we can add the setting to the installation page, but it doesn't work with USO styles.

**Comment by eight04** [comment]
Some screenshots:
image
image

**Comment by kristofferR** [comment]
Thanks, but it doesn't detect Dark Mode properly yet. Nothing happens in Dark Mode:
https://streamable.com/12e1i

As you can see I also tested a normal website (webkit.org) with a native dark css, so the problem is in Stylus somewhere.

**Comment by eight04** [comment]
Unfortunately, I'm using Windows 7 so I don't know how to test system integration.

Can you try restarting the browser and see if the style is applied? If the extension can get the theme state on startup but fail to update it when the system theme is changed, I guess the event is not fired correctly in the background:
https://github.com/openstyles/stylus/blob/ce55c89fbf7177891b8a9163c9c3d5127660b4eb/background/color-scheme.js#L12

**Comment by eight04** [comment]
Here is the Chrome bug: https://crbug.com/968651

**Comment by kristofferR** [comment]
I tried it now and yeah, it activates if Chrome was launched during dark mode. I'll try Firefox soon.

**Comment by gdh1995** [comment]
In fact, during my tests on Firefox 67.0.1, the listener is also not triggerred, and what's worse is Firefox won't  update the `media.matches`. So currently, I just set an interval to query the medias every 30 seconds.

**Comment by eight04** [comment]
Maybe relate: https://bugzilla.mozilla.org/show_bug.cgi?id=1561546

File: popup/popup.js [hunk 1]
```diff
@@ -343,6 +344,10 @@ function createStyleElement(style) {
   return entry;
 }
 
+function capitalize(s) {
+  return s[0].toUpperCase() + s.slice(1);
+}
+
 function styleExcluded({exclusions}, type) {
```
Comment (Mottie) [inline popup/popup.js]: I don't think we should manipulate translations like this... even using [`text-transform`](https://developer.mozilla.org/en-US/docs/Mozilla/Localization/Localization_content_best_practices#CSS_issues) isn't reliable.

**Comment by eight04** [comment]
> * CodeMirror takes a very long time to load GitHub-Dark. It's a difference of **2-3 seconds** when I compare this branch to the master.

CodeMirror? Where?

> * I think each individual style should have toggle light/dark mode settings that override the advanced option setting. We could add this as a popup on the manager page or maybe even the edit page. I'll have to think about how to manage multiple styles in the new manager PR.

Can you elaborate? Do you mean to allow users to set different night time for each style?

> * Would there ever be the case where a user would want to add more than one time period to toggle light/dark more?

I'm not familiar with dark (night) mode apps. Is there any other application which can configure multiple periods?

File: popup/popup.js (hunk 1 above)
Comment (eight04) [reply popup/popup.js]: It doesn't manipulate translations. It is just a helper function to capitalize the first character. For example:
`'installPreferScheme' + 'dark'` -> `'installPreferScheme' + 'Dark'`

Comment (Mottie) [reply popup/popup.js]: Oh ok... I missed that because I was lacking caffeine 😸

**Comment by Mottie** [comment]
> CodeMirror? Where?

I don't see any code that modifies CodeMirror in this PR, but I was just making an observation that Opening up GitHub-Dark in the editor from the popup or the manager is very slow.

> Can you elaborate? Do you mean to allow users to set different night time for each style?

I think that might be desirable. In the original PR every style had it's own time setting.

> Is there any other application which can configure multiple periods?

Not that I know of, I was asking as a more general question. @narcolepticinsomniac WDYT?

**Comment by eight04** [comment]
> I don't see any code that modifies CodeMirror in this PR, but I was just making an observation that opening up GitHub-Dark in the editor from the popup or the manager is very slow.

It seems that CodeMirror is updated on master: b7b6ed2ff5eb69a937e020581fd89b513664c694. Maybe the performance is improved?

**Comment by Mottie** [comment]
Yes, merging master into this branch does appear to fix the performance.

**Comment by Cimbali** [comment]
Maybe the choice of what modes (night/day/always) a stylesheet is applied to could appear/be modified on its “Edit style” page, rather than just the install page?

**Comment by eight04** [comment]
> Maybe the choice of what modes (night/day/always) a stylesheet is applied to could appear/be modified on its “Edit style” page, rather than just the install page?

We will probably create a style-setting page in the future and put the option there. Including this PR, update URL, user-defined includes/excludes, etc.

**Comment by kristofferR** [comment]
Dark Reader was able to bypass the Chrome and Firefox bugs:
https://github.com/darkreader/darkreader/pull/1522

They've shipped a new release with Dark Mode support, and it works great on both Firefox and Chrome. You might want to take a look to see how they've solved it.

**Comment by eight04** [comment]
They observe the change in the content script, which is not affected by the background page bugs. I think we can also use this workaround. However, since it depends on content scripts, it requires at least an opened/activated page. Not sure what will happen when opening a new page while no content script is available.

**Comment by eight04** [comment]
This branch now detects the media change in the content script. Please test it again.

**Comment by gdh1995** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1561546 is marked fixed, but I can not get `onchange` triggered on Firefox 71 + Win 10, so I just create a new issue https://bugzilla.mozilla.org/show_bug.cgi?id=1587723

**Comment by kristofferR** [comment]
It works great now. :D

The user experience is a horrible mess though since only remote userCSSs are supported of course, so whatever styles I find I need to copy and rehost and install, but I see that you're aware of that issue and working on fixing the UI (although it seems both UI PRs have become stale lately).

**Comment by LirSegev** [comment]
> It works great now. :D
> 
> The user experience is a horrible mess though since only remote userCSSs are supported of course, so for every style I find I need to manually copy it, add the usercss header, rehost the style and install it, but I see that you're aware of that issue and working on fixing the UI (although it seems both UI PRs have become stale lately).
> 
> Perhaps a temp fix would be to add install settings support for locally created userCSSs? That'll remove the rehosting step at least.

How are you doing this? I can't get this to work.

**Comment by kristofferR** [comment]
> > It works great now. :D
> > The user experience is a horrible mess though since only remote userCSSs are supported of course, so for every style I find I need to manually copy it, add the usercss header, rehost the style and install it, but I see that you're aware of that issue and working on fixing the UI (although it seems both UI PRs have become stale lately).
> 
> @kristofferR How are you doing this? I can't get this to work.

You need to do both things in the screenshots here (chose "By system preference" in screenshot 2):
https://github.com/openstyles/stylus/pull/736#issuecomment-505745358

**Comment by patroza** [comment]
This works like a dream! Thanks so much.
Tested on latest Chromium Edge and Chrome, MacOSX Catalina.

**Comment by Logarithmus** [comment]
I've downloaded `dev-color-scheme` branch, there's new option from 2nd screenshot, https://github.com/openstyles/stylus/pull/736#issuecomment-505745358, but I couldn't find the option from the 1st screenshot. What's wrong? Please help. I'd really love to automate light/dark styles switching, it's so annoying to do it manually twice a day...

**Comment by dnknn** [comment]
If want to do it, The most perfect and convenient should be like this :▼
Snipaste_2020-06-01_17-34-20

Users can easily ON/OFF at any time without any difficulty.
`#stylus-options` not used for switching, but for advanced settings, such as Toggle Light/Dark Mode automatically.

WDYT？

**Comment by ShivamJoker** [comment]
I don't see any option to enable only in dark mode 
image

**Comment by dnknn** [comment]
> I don't see any option to enable only in dark mode

Don't worry, it's not realized yet.
Dark mode is not the most important.

**Comment by ShivamJoker** [comment]
> > I don't see any option to enable only in dark mode
> 
> Don't worry, it's not realized yet.
> Dark mode is not the most important.

But I saw screenshot in the thread thats why i am asking how can I get that working so that I don't have to manually turn it on and off ~_~

**Comment by narcolepticinsomniac** [comment]
> I saw screenshot in the thread

https://user-images.githubusercontent.com/1324510/60157672-e7b12600-9821-11e9-8513-d391f5aa6a54.png

If you're referring to this one, that's a usercss install page. I guess maybe that's how this PR is configured currently? IDK, I've never tested it.

IIRC, the general plan for the UI was to eventually include these settings in config menus, along with some other per-style settings, but there hasn't been much progress in that direction so far.

**Comment by dnknn** [comment]
I can't see too

<details>
<summary>pic</summary>

image

</details>

**Comment by narcolepticinsomniac** [comment]
>  maybe that's how this PR is configured currently?

¯\_(ツ)_/¯ Wild guess based off the screenshot. Like I said, I never tested, nor do I plan to in its current state. Maybe you need to reinstall the style? Maybe it doesn't work at all? Who knows?

**Comment by michaelgosling** [comment]
> > I don't see any option to enable only in dark mode
> 
> Don't worry, it's not realized yet.
> Dark mode is not the most important.

FWIW, I think respecting system color scheme should be a major priority for an extension that's primary objective is to provide custom styling to users. Having to edit and wrap every update of a userstyle in media queries is tiresome. This is a major feature of [Cascadea](https://cascadea.app/) for Safari, and even VSCode has support for this. I would say demand (and therefore importance) is rather high, IMO.

**Comment by eight04** [comment]
> If want to do it, The most perfect and convenient should be like this :▼
> Snipaste_2020-06-01_17-34-20
> 
> Users can easily ON/OFF (Auto toggle Light/Dark) at any time without any difficulty.
> `#stylus-options` not used for switching, but for advanced settings, such as Toggle Light/Dark Mode automatically.
> 
> WDYT？

I think it makes more sense to set the option to "By system preference" and flip the mode from system UI.

**Comment by dnknn** [comment]
> makes more sense to set the option to "By system preference" and flip the mode from system UI.

What that means? sorry for my English..

**Comment by michaelgosling** [comment]
> > makes more sense to set the option to "By system preference" and flip the mode from system UI.
> 
> What that means? sorry for my English..
> 
> mean : gif-demo ?
> [darkreader/darkreader#1522](https://github.com/darkreader/darkreader/pull/1522)
> image

I believe thats exactly what they mean, yes.

**Comment by dnknn** [comment]
ms-settings:nightlight

**Comment by eddie-LA** [comment]
Hello. 
I was going to file an issue but then found this.
Any progress with this..? I also use Stylish mainly for the more efficient Dark Mode. 😄

**Comment by eight04** [comment]
The core feature is ready. However, currently you can only configure this option for usercss *at installation time*.

The configuration UI is blocked by #827.

**Comment by eight04** [comment]
> I don't know what problems would ensue due to merging it now

If we drop a release before fixing #827, we will probably receive some issues about not being able to configure dark/light mode for some styles. To avoid that, we can hide the UI via CSS so those advanced users can still access the options by showing them with custom CSS.

**Comment by Draculah** [comment]
Just like #1343, I was wondering when/if the auto per-theme toggle based on system preferences would show be available. Is this the latest update on the auto update theme?

And, assuming it's in the extension by the merge, what CSS would I need to display these settings? Or is it highly _not_ recommended?

**Comment by tophf** [comment]
It's in the beta.

**Pull request merged**